pennylane/devices/qubit/measure.py

```diff
@@ -71,56 +71,57 @@ def state_hamiltonian_expval(measurementprocess: ExpectationMP, state: TensorLik
+        state (TensorLike): the state to measure
 
     Returns:
         TensorLike: the result of the measurement
     """
-    total_indices = len(state.shape)
-    wires = Wires(range(total_indices))
-    return measurementprocess.process_state(math.flatten(state), wires)
+    if isinstance(measurementprocess.obs, Sum):
+        # Recursively call measure on each term, so that the best measurement method can
+        # be used for each term
+        return sum(measure(ExpectationMP(term), state) for term in measurementprocess.obs)
+    # else hamiltonian
+    return sum(
+        c * measure(ExpectationMP(t), state) for c, t in zip(*measurementprocess.obs.terms())
+    )
 
 
 def get_measurement_function(
-    measurementprocess: MeasurementProcess,
+    measurementprocess: MeasurementProcess, state: TensorLike
 ) -> Callable[[MeasurementProcess, TensorLike], TensorLike]:
     """Get the appropriate method for performing a measurement.
 
     Args:
         measurementprocess (MeasurementProcess): measurement process to apply to the state
+        state (TensorLike): the state to measure
 
     Returns:
         Callable: function that returns the measurement result
     """
     if isinstance(measurementprocess, StateMeasurement):
-        if measurementprocess.obs is None:
-            # no need to apply diagonalizing gates
-            return state_measurement_process
-
-        if isinstance(measurementprocess, ExpectationMP) and measurementprocess.obs.name in (
-            "Hamiltonian",
-            "SparseHamiltonian",
-        ):
-            return state_hamiltonian_expval
-
-        if (
-            isinstance(measurementprocess, ExpectationMP)
-            and isinstance(measurementprocess.obs, Sum)
-            and measurementprocess.obs.has_overlapping_wires
-            and len(measurementprocess.obs.wires) > 7
-        ):
-            # Use tensor contraction for `Sum` expectation values with non-commuting summands
-            # and 8 or more wires as it's faster than using eigenvalues.
-            return state_hamiltonian_expval
-
-        if measurementprocess.obs.has_diagonalizing_gates:
+        if isinstance(measurementprocess, ExpectationMP):
+            if measurementprocess.obs.name == "SparseHamiltonian":
+                return csr_dot_products
+
+            if measurementprocess.obs.name == "Hamiltonian" or (
+                isinstance(measurementprocess.obs, Sum)
+                and measurementprocess.obs.has_overlapping_wires
+                and len(measurementprocess.obs.wires) > 7
+            ):
+                # Use tensor contraction for `Sum` expectation values with non-commuting summands
+                # and 8 or more wires as it's faster than using eigenvalues.
+
+                # need to work out thresholds for when its faster to use "backprop mode" measurements
+                backprop_mode = math.get_interface(state) != "numpy"
```

Comment:
I chose to pass `state` to `get_measurement_function` instead so that it would match the call signature of all the other methods.

Also, we would only need to calculate `backprop_mode` when we are actually going to use it. I'm not too concerned about the name of the variable as it's immediately consumed.

Reply:
reasoning all sounds good to me!